Replace stdlib write/read with send/recv (Fixes #12673)

franz1981 opened the pull request:
Motivation:

The performance Unix write/read paths is more involved (and slower) then the specialized socket send/rcv ones.

Modification:

Replace Unix write/read paths with send/recv

Result:

Better performance for single buffer send/recv

franz1981 commented:
@normanmaurer waiting the CI :)

normanmaurer commented on transport-native-unix-common/src/main/c/netty_unix_filedescriptor.c:
```diff
@@ -41,7 +42,7 @@ static jint _write(JNIEnv* env, jclass clazz, jint fd, void* buffer, jint pos, j
     do {
-       res = write(fd, buffer + pos, (size_t) (limit - pos));
+       res = send(fd, buffer + pos, (size_t) (limit - pos), 0);
        // keep on writing if it was interrupted
```
I think it would be better to add extra method for this and not just replace the existing ones

franz1981 replied:
turning into draft and fixing this, come back soon :)

franz1981 commented:
@normanmaurer comments are welcome, undrafting

normanmaurer commented on transport-native-unix-common/src/main/c/netty_unix_filedescriptor.c:
```diff
@@ -69,7 +70,7 @@ static jint _read(JNIEnv* env, jclass clazz, jint fd, void* buffer, jint pos, ji
     do {
-        res = read(fd, buffer + pos, (size_t) (limit - pos));
+        res = recv(fd, buffer + pos, (size_t) (limit - pos), 0);
         // Keep on reading if we was interrupted
```
Revert changes here ?

normanmaurer commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/LinuxSocket.java:
```diff
@@ -300,6 +301,44 @@ void setUdpGro(boolean gro) throws IOException {
+        if (res == 0) {
+            return -1;
+        }
+        return ioResult("read", res);
```
```suggestion
        return ioResult("recv", res);
```

normanmaurer commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/LinuxSocket.java:
```diff
@@ -300,6 +301,44 @@ void setUdpGro(boolean gro) throws IOException {
+        if (res == 0) {
+            return -1;
+        }
+        return ioResult("readAddress", res);
```
```suggestion
        return ioResult("recvAddress", res);
```

normanmaurer commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/LinuxSocket.java:
```diff
@@ -300,6 +301,44 @@ void setUdpGro(boolean gro) throws IOException {
+        if (res >= 0) {
+            return res;
+        }
+        return ioResult("write", res);
```
```suggestion
        return ioResult("send", res);
```

normanmaurer commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/LinuxSocket.java:
```diff
@@ -300,6 +301,44 @@ void setUdpGro(boolean gro) throws IOException {
+        if (res >= 0) {
+            return res;
+        }
+        return ioResult("writeAddress", res);
```
```suggestion
        return ioResult("sendAddress", res);
```

franz1981 commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java:
```diff
@@ -1043,6 +1043,12 @@ public boolean spliceIn(RecvByteBufAllocator.Handle handle) {
+
+        @Override
+        protected int write(final ByteBuffer buf, final int pos, final int limit) throws IOException {
+            return socket.send(buf, pos, limit);
```
That's an hack because `SocketWritableByteChannel' is a public API and I cannot introduce a generic param `T extends FileDescriptor` on it @normanmaurer

franz1981 replied:
I really don't like this trick @normanmaurer :( I feel dirty

franz1981 commented:
I've added some `protected final` method to `Socket` and the checker complain I've modified a public API :(
I'm forced to move send/recv in the epoll socket c file @normanmaurer ?
I would like to keep it on unix socket, in case we want to port this to kqueue (after being tested)

normanmaurer commented:
> I've added some `protected final` method to `Socket` and the checker complain I've modified a public API :(
> I'm forced to move send/recv in the epoll socket c file @normanmaurer ?
> I would like to keep it on unix socket, in case we want to port this to kqueue (after being tested)

just remove final

normanmaurer commented:
@franz1981 let me know once I should check again

franz1981 commented:
@normanmaurer fingers crossed!
I'll run some bench on our lab in the afternoon (although using the full quarkus stack ^^)

normanmaurer commented:
@franz1981 please add the numbers here

franz1981 commented:
The firsts load tests using just the receive side (ie read vs recv) shows this improvement:
~4_200_000 requests/sec vs ~4_700_000 requests/sec

The test used has been Techempower plaintext with quarkus (meaning a full-stack, not just netty) using
```
wrk -H 'Host: 192.168.0.106' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: keep-alive' --latency -d 15 -c 256 --timeout 8 -t 64 http://192.168.0.106:8080/plaintext -s pipeline.lua -- 16
```
vs the full reactive stack.
Due to pipelining, the write side is likely using `writev` heance the improvement is just due to the recv side change

normanmaurer commented:
Also /cc @Lukasa for visibility

normanmaurer commented:
/cc @trustin

Lukasa commented:
Interesting, where's the data coming from to suggest that `send` outperforms `write`? Relatedly, does the same thing apply to `sendmsg` outperforming `writev`?

normanmaurer commented:
@Lukasa the issue has some more details as well https://github.com/netty/netty/issues/12673

franz1981 commented:
Consider @Lukasa that the most benefit come by the different specialization of the syscalls: send/recv are indeed for socket fds, while write/read can be used for other types of fds
The mentioned libreactor issue better explain the involved kernel code paths

franz1981 commented:
@normanmaurer given I've modified a socket public API I'm prepared to some failure :(
but the comments are addressed :)

franz1981 commented:
@normanmaurer re https://github.com/netty/netty/pull/12679#issuecomment-1221309563 I was wrong, so, ready to go!

normanmaurer commented:
Thanks a lot!
